client/consensus/grandpa/environment.go

```diff
@@ -0,0 +1,466 @@
+	}
+	tve := VoterSetState[H, N](vdt)
+	return &tve
+}
+
+// Live Create a new live VoterSetState with round 0 as a completed round using
+// the given genesis state and the given authorities. Round 1 is added as a
+// current round (with state `HasVoted::No`).
+func (tve *VoterSetState[H, N]) Live(setId uint64, authSet AuthoritySet[H, N], genesisState grandpa.HashNumber[H, N]) (VoterSetState[H, N], error) {
+	state := grandpa.NewRoundState[H, N](genesisState)
+	completedRounds := NewCompletedRounds[H, N](
+		&CompletedRound[H, N]{
+			State: state,
+			Base:  genesisState,
+		},
+		setId,
+		authSet,
+	)
+	currentRounds := CurrentRounds[H, N](
+		make(map[uint64]HasVoted[H, N]),
+	)
+	hasVoted := HasVoted[H, N]{}
+	hasVoted = hasVoted.New()
+	err := hasVoted.Set(No{})
+	if err != nil {
+		return VoterSetState[H, N]{}, err
+	}
+	currentRounds[1] = hasVoted
+
+	voterSetState := *NewVoterSetState[H, N]()
+	err = voterSetState.Set(Live[H, N]{
+		CompletedRounds: completedRounds,
+		CurrentRounds:   currentRounds,
+	})
+
+	if err != nil {
+		return VoterSetState[H, N]{}, err
+	}
+	return voterSetState, nil
+}
+
+// CompletedRounds Returns the last completed rounds
+func (tve *VoterSetState[H, N]) CompletedRounds() (CompletedRounds[H, N], error) {
+	value, err := tve.Value()
+	if err != nil {
+		return CompletedRounds[H, N]{}, err
+	}
+	switch v := value.(type) {
+	case Live[H, N]:
+		return v.CompletedRounds, nil
+	case Paused[H, N]:
+		return v.CompletedRounds, nil
+	default:
+		panic("CompletedRounds: invalid voter set state")
+	}
+}
+
+// LastCompletedRound Returns the last completed round
+func (tve *VoterSetState[H, N]) LastCompletedRound() (CompletedRound[H, N], error) {
+	value, err := tve.Value()
+	if err != nil {
+		return CompletedRound[H, N]{}, err
+	}
+	switch v := value.(type) {
+	case Live[H, N]:
+		return v.CompletedRounds.Last(), nil
+	case Paused[H, N]:
+		return v.CompletedRounds.Last(), nil
+	default:
+		panic("CompletedRounds: invalid voter set state")
+	}
+}
+
+// WithCurrentRound Returns the voter set state validating that it includes the given round
+// in current rounds and that the voter isn't paused
+func (tve *VoterSetState[H, N]) WithCurrentRound(round uint64) (CompletedRounds[H, N], CurrentRounds[H, N], error) {
+	value, err := tve.Value()
+	if err != nil {
+		return CompletedRounds[H, N]{}, CurrentRounds[H, N]{}, err
+	}
+	switch v := value.(type) {
+	case Live[H, N]:
+		_, contains := v.CurrentRounds[round]
+		if contains {
+			return v.CompletedRounds, v.CurrentRounds, nil
+		}
+		return CompletedRounds[H, N]{}, CurrentRounds[H, N]{}, fmt.Errorf("voter acting on a live round we are not tracking")
+	case Paused[H, N]:
+		return CompletedRounds[H, N]{}, CurrentRounds[H, N]{}, fmt.Errorf("voter acting while in paused state")
+	default:
+		panic("CompletedRounds: invalid voter set state")
+	}
+}
+
+// Live The voter is live, i.e. participating in rounds.
+type Live[H comparable, N constraints.Unsigned] struct {
```

Comment:
maybe more explicit to what this is a `VaryingDataTypeValue` for.
```suggestion
type voterSetStateLive[H comparable, N constraints.Unsigned] struct {
```